alerting/src/test/kotlin/org/opensearch/alerting/bwc/AlertingBackwardsCompatibilityIT.kt

```diff
@@ -0,0 +1,220 @@
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.alerting.bwc
+
+import org.apache.http.entity.ContentType.APPLICATION_JSON
+import org.apache.http.entity.StringEntity
+import org.opensearch.alerting.ALERTING_BASE_URI
+import org.opensearch.alerting.AlertingRestTestCase
+import org.opensearch.alerting.LEGACY_OPENDISTRO_ALERTING_BASE_URI
+import org.opensearch.alerting.makeRequest
+import org.opensearch.alerting.model.Monitor
+import org.opensearch.common.settings.Settings
+import org.opensearch.common.xcontent.XContentType
+import org.opensearch.index.query.QueryBuilders
+import org.opensearch.rest.RestStatus
+import org.opensearch.search.builder.SearchSourceBuilder
+
+class AlertingBackwardsCompatibilityIT : AlertingRestTestCase() {
+
+    companion object {
+        private val CLUSTER_TYPE = ClusterType.parse(System.getProperty("tests.rest.bwcsuite"))
+        private val CLUSTER_NAME = System.getProperty("tests.clustername")
+    }
+
+    override fun preserveIndicesUponCompletion(): Boolean = true
+
+    override fun preserveReposUponCompletion(): Boolean = true
+
+    override fun preserveTemplatesUponCompletion(): Boolean = true
+
+    override fun preserveODFEIndicesAfterTest(): Boolean = true
+
+    override fun restClientSettings(): Settings {
+        return Settings.builder()
+            .put(super.restClientSettings())
+            // increase the timeout here to 90 seconds to handle long waits for a green
+            // cluster health. the waits for green need to be longer than a minute to
+            // account for delayed shards
+            .put(CLIENT_SOCKET_TIMEOUT, "90s")
+            .build()
+    }
+
+    @Throws(Exception::class)
+    @Suppress("UNCHECKED_CAST")
+    fun `test backwards compatibility`() {
+        val uri = getPluginUri()
+        val responseMap = getAsMap(uri)["nodes"] as Map<String, Map<String, Any>>
+        for (response in responseMap.values) {
+            val plugins = response["plugins"] as List<Map<String, Any>>
+            val pluginNames = plugins.map { plugin -> plugin["name"] }.toSet()
+            when (CLUSTER_TYPE) {
+                ClusterType.OLD -> {
+                    assertTrue(pluginNames.contains("opendistro-alerting"))
+                    createBasicMonitor()
+                }
+                ClusterType.MIXED -> {
+                    assertTrue(pluginNames.contains("opensearch-alerting"))
+                    verifyMonitorExists(LEGACY_OPENDISTRO_ALERTING_BASE_URI)
+                    // Waiting a minute to ensure the Monitor ran again at least once before checking if the job is running
+                    // on time
+                    // TODO: Should probably change the next execution time of the Monitor manually instead since this inflates
+                    //  the test execution by a lot
+                    Thread.sleep(60000)
```

Comment:
Can we avoid sleep(...) ?

Reply:
I had originally planned on adding a utility method like what ISM had added:

https://github.com/opensearch-project/index-management/blob/ab122799283a58bb3f5c6f6d8576c7ba5875f13b/src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/IndexStateManagementRestTestCase.kt#L382

As well as a `waitFor` wrapper so we could add assertions that could wait for the Monitor to execute and confirm state afterwards:

https://github.com/opensearch-project/index-management/blob/ab122799283a58bb3f5c6f6d8576c7ba5875f13b/src/test/kotlin/org/opensearch/indexmanagement/TestHelpers.kt#L133-L149

However, it looks like the JobScheduler implementation of the schedule information stores things like `start_time` which is convenient to modify for the particular scheduled job you'd like to speed up in test scenarios:

https://github.com/opensearch-project/job-scheduler/blob/main/spi/src/main/java/org/opensearch/jobscheduler/spi/schedule/IntervalSchedule.java#L54

Alerting's scheduled job implementation does not do this and instead uses `enabled_time` to resolve the next expected execution time only if the `expectedPreviousExecutionTime` is not set, otherwise it uses that:

https://github.com/opensearch-project/alerting/blob/e0e41df2b2e36c187edd3c45e214451208b10309/core/src/main/kotlin/org/opensearch/alerting/core/model/Schedule.kt#L332

The `expectedPreviousExecutionTime` is stored in `scheduledJobIdToInfo` which is part of a concurrent hashmap which is stored in memory as far as I'm aware:
https://github.com/opensearch-project/alerting/blob/e0e41df2b2e36c187edd3c45e214451208b10309/core/src/main/kotlin/org/opensearch/alerting/core/schedule/JobScheduler.kt#L52

This would make modification of an already running Monitor to execute sooner less trivial than expected. I'd like to leave it out of scope of this particular PR for now and have it be a possible follow-up for a test refactor.

Reply:
Hmm, looks like these tests do get run automatically as part of `./gradlew build` so this would impact build times when running all tests. I was hoping to exclude them unless running something like `./gradlew bwcTestSuite` to avoid the overhead for the time being but `./gradlew build` will likely pick up all tests by nature.

Reply:
Thanks for diving deep into this. Currently `gradlew build` takes around 4mins, this sleep would add one more min.

Reply:
Actually, since the `bwcTestSuite` executes the following scenarios:

1. `mixedCluster` - Hits the `MIXED` case
2. `rollingUpgrade` - Hits the `MIXED` and `UPGRADED` case
3. `fullRestart` - Hits the `UPGRADED` case

We'd be adding 4 minutes of sleep across all the tests since both the `MIXED` and `UPGRADED` cases are adding a sleep of 1 minute before verifying.